Announce Rails 7 alpha 1 (#260)
rails/weblog: _posts/2021-09-15-Rails-7-0-alpha-1-released.markdown

---
layout: post
title: 'Rails 7.0 Alpha 1: New JavaScript Answers, At-Work Encryption, Query Origin Logging, Zeitwerk Exclusively'
categories: releases
author: dhh
published: true
date: 2021-09-15 14:00:00 -08:00
---
Welcome to the first alpha release of Rails 7. It brings some very exciting new answers to how we do JavaScript, an awesome approach to at-work encryption with Active Record, SQL query origin logging, exclusive autoloading through Zeitwerk, and much more.

We usually don't do alpha releases for Rails, but given the fact that the new front-end approach is such a substantial change, we thought it best to validate that a little further before jumping straight on the beta -> release candidate -> final train.

Please help us test all this new stuff so we can ensure a solid final release of Rails 7 this year!

## All New Answers On The Front-End

> After almost five years with Webpacker as our default answer to writing modern JavaScript in Rails, [it's time to move on](https://www.youtube.com/watch?v=PtxZvFnL2i0). Advancements in browser support for ES6/ESM, widespread adoption of HTTP2, and [the exciting new standard for import maps](https://github.com/rails/importmap-rails) has [paved the way](https://world.hey.com/dhh/modern-web-apps-without-javascript-bundling-or-transpiling-a20f2755) for [a no-Node approach to JavaScript in Rails 7](https://world.hey.com/dhh/rails-7-will-have-three-great-answers-to-javascript-in-2021-8d68191b) – [without giving up on npm packages](https://github.com/rails/importmap-rails#using-node-modules-via-javascript-cdns).

Reviewer: ```suggestion
After almost five years with [Webpacker](https://github.com/rails/webpacker) as our default answer to writing modern JavaScript in Rails, [it's time to move on](https://www.youtube.com/watch?v=PtxZvFnL2i0). Advancements in browser support for ES6/ESM, widespread adoption of HTTP2, and [the exciting new standard for import maps](https://github.com/rails/importmap-rails) has [paved the way](https://world.hey.com/dhh/modern-web-apps-without-javascript-bundling-or-transpiling-a20f2755) for [a no-Node approach to JavaScript in Rails 7](https://world.hey.com/dhh/rails-7-will-have-three-great-answers-to-javascript-in-2021-8d68191b) – [without giving up on npm packages](https://github.com/rails/importmap-rails#using-node-modules-via-javascript-cdns).
```

> Together with the replacement of Turbolinks and Rails UJS by the [Hotwire](https://hotwired.dev) combination of [Stimulus](https://stimulus.hotwired.dev) and [Turbo](https://turbo.hotwired.dev), we now have the most complete in-the-box front-end setup for writing great Rails applications ever. Without needing thousands of node dependencies in node_modules, fighting with bundler configurations, or any of the other challenges common with JavaScript development.

Reviewer: ```suggestion
Together with the replacement of Turbolinks and Rails UJS by the [Hotwire](https://hotwired.dev) combination of [Stimulus](https://stimulus.hotwired.dev) and [Turbo](https://turbo.hotwired.dev), we now have the most complete in-the-box front-end setup for writing great Rails applications ever. Without needing thousands of node dependencies in `node_modules`, fighting with bundler configurations, or any of the other challenges common with JavaScript development.
```

> At the same time, we've also [dramatically improved the integration](https://www.youtube.com/watch?v=JsNtLiph87Y) between Rails and JavaScript + CSS bundlers for those who need that. Through two new companion gems that can be triggered via rails new [--javascript [bundler]](https://github.com/rails/jsbundling-rails/) and [--css [bundler]](https://github.com/rails/cssbundling-rails/), you get easy access to starting a new application or changing one that starts with import maps to use esbuild, rollup.js, Webpack, Tailwind CSS, PostCSS, Dart Sass, and Bootstrap.

Reviewer: ```suggestion
At the same time, we've also [dramatically improved the integration](https://www.youtube.com/watch?v=JsNtLiph87Y) between Rails and JavaScript + CSS bundlers for those who need that. Through two new companion gems that can be triggered via rails new [--javascript [bundler]](https://github.com/rails/jsbundling-rails/) and [--css [bundler]](https://github.com/rails/cssbundling-rails/), you get easy access to starting a new application or changing one that starts with import maps to use [esbuild](https://github.com/evanw/esbuild), [rollup.js](https://rollupjs.org/guide/en/), [Webpack](https://webpack.js.org/), [Tailwind CSS](https://tailwindcss.com/), [PostCSS](https://postcss.org/), [Dart Sass](https://sass-lang.com/dart-sass), and [Bootstrap](https://getbootstrap.com/).
```

As an immediate practical benefit, encrypting sensitive attributes adds an additional security layer. For example, if an attacker gained access to your database, a snapshot of it, or your application logs, they wouldn't be able to make sense of the encrypted information. And even without thinking about malicious actors, checking application logs for legit reasons shouldn't expose personal information from customers either.

But more importantly, by using Active Record Encryption, you define what constitutes sensitive information in your application at the code level. This enables controlling how this information is accessed and building services around it. As examples, think about auditable Rails consoles that protect encrypted data or check the built-in system to filter controller params automatically.

Checkout the [full guide on how to use encrypted attributes](https://edgeguides.rubyonrails.org/active_record_encryption.html).

> ## Trace query origins with Marginalia-style tagging

Reviewer: ```suggestion
## Trace Query Origins With Marginalia-Style Tagging
```

Almost a decade ago, [Marginalia was extracted from Basecamp](https://signalvnoise.com/posts/3130-tech-note-mysql-query-comments-in-rails) to trace query origins with SQL comment tagging. Now this external gem has been [upstreamed into Active Record as QueryLogs](https://github.com/rails/rails/pull/42240).

## Zeitwerk Exclusively

> Autoloading in Rails is one of those magical quality of life realities that its easy to just take for granted. The trusty old const_missing approach which came with a range of quirks and missing features has finally been replaced exclusively with the [Zeitwerk code loader](https://github.com/fxn/zeitwerk#introduction). There are a few upgrade gotchas to be aware of, especially for older applications, but with this [upgrade guide](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#autoloading) you should be on your way in no time

Reviewer: ```suggestion
Autoloading in Rails is one of those magical quality of life realities that it's easy to just take for granted. The trusty old `const_missing` approach which came with a range of quirks and missing features has finally been replaced exclusively with the [Zeitwerk code loader](https://github.com/fxn/zeitwerk#introduction). There are a few upgrade gotchas to be aware of, especially for older applications, but with this [upgrade guide](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#autoloading) you should be on your way in no time
```

> ## A few other highlights

Reviewer: ```suggestion
## A Few Other Highlights
```

> - Spring is no longer on by default, as faster computers have made it less relevant on anything but the largest applications.

Reviewer: ```suggestion
- [Spring](https://github.com/rails/spring) is no longer on by default, as faster computers have made it less relevant on anything but the largest applications.
```

- [ActionController::Live#send_stream](https://github.com/rails/rails/pull/41488) makes it easy to stream files that are being generated on-the-fly in controller actions.
- Parallelized testing will now compare your CPU core count to your test count and [scale the parallelization accordingly](https://github.com/rails/rails/pull/42761).

> ## From all of us to all of you

Reviewer: ```suggestion
## From All Of Us To All Of You
```

---
layout: post
title: 'Rails 7.0 Alpha 1: New JavaScript Answers, At-Work Encryption, Query Origin Logging, Zeitwerk Exclusively'
categories: releases
author: dhh
published: true
date: 2021-09-15 14:00:00 -08:00
---
Welcome to the first alpha release of Rails 7. It brings some very exciting new answers to how we do JavaScript, an awesome approach to at-work encryption with Active Record, SQL query origin logging, asynchronous query loading, exclusive autoloading through Zeitwerk, and much more.

We usually don't do alpha releases for Rails, but given the fact that the new front-end approach is such a substantial change, we thought it best to validate that a little further before jumping straight on the beta -> release candidate -> final train.

Please help us test all this new stuff so we can ensure a solid final release of Rails 7 this year!

## All New Answers On The Front-End

> After almost five years with [Webpacker](https://github.com/rails/webpacker) as our default answer to writing modern JavaScript in Rails, [it's time to move on](https://www.youtube.com/watch?v=PtxZvFnL2i0). Advancements in browser support for ES6/ESM, widespread adoption of HTTP2, and [the exciting new standard for import maps](https://github.com/rails/importmap-rails) has [paved the way](https://world.hey.com/dhh/modern-web-apps-without-javascript-bundling-or-transpiling-a20f2755) for [a no-Node approach to JavaScript in Rails 7](https://world.hey.com/dhh/rails-7-will-have-three-great-answers-to-javascript-in-2021-8d68191b) – [without giving up on npm packages](https://github.com/rails/importmap-rails#using-node-modules-via-javascript-cdns).

Reviewer 2: Minor: https://en.wikipedia.org/wiki/HTTP/2 

```suggestion
After almost five years with [Webpacker](https://github.com/rails/webpacker) as our default answer to writing modern JavaScript in Rails, [it's time to move on](https://www.youtube.com/watch?v=PtxZvFnL2i0). Advancements in browser support for ES6/ESM, widespread adoption of HTTP/2, and [the exciting new standard for import maps](https://github.com/rails/importmap-rails) has [paved the way](https://world.hey.com/dhh/modern-web-apps-without-javascript-bundling-or-transpiling-a20f2755) for [a no-Node approach to JavaScript in Rails 7](https://world.hey.com/dhh/rails-7-will-have-three-great-answers-to-javascript-in-2021-8d68191b) – [without giving up on npm packages](https://github.com/rails/importmap-rails#using-node-modules-via-javascript-cdns).
```

## Zeitwerk Exclusively

Autoloading in Rails is one of those magical quality of life realities that it's easy to just take for granted. The trusty old `const_missing` approach which came with a range of quirks and missing features has finally been replaced exclusively with the [Zeitwerk code loader](https://github.com/fxn/zeitwerk#introduction). There are a few upgrade gotchas to be aware of, especially for older applications, but with this [upgrade guide](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#autoloading) you should be on your way in no time

## A Few Other Highlights

- [Spring](https://github.com/rails/spring) is no longer on by default, as faster computers have made it less relevant on anything but the largest applications.
- [ActionController::Live#send_stream](https://github.com/rails/rails/pull/41488) makes it easy to stream files that are being generated on-the-fly in controller actions.

> - Parallelized testing will now compare your CPU core count to your test count and [scale the parallelization accordingly](https://github.com/rails/rails/pull/42761).

Reviewer 3: ```suggestion
- Parallelized testing will now compare your CPU core count to your test count and [scale the parallelization accordingly](https://github.com/rails/rails/pull/42761).
- Active Storage now uses the faster and more secure `libvips` as its [default variant processor](https://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#activestorage-variant-processor-changed-to-vips).
```

Not sure what the criteria for highlights is, but I think this is a neat upgrade.
